Add call to animate model to center of portal on drag

Phipson opened the pull request:
#### ed9d5b3d67342786a842d7613a5088281fae3bfa
<pre>
Add call to animate model to center of portal on drag
<a href="https://bugs.webkit.org/show_bug.cgi?id=291288">https://bugs.webkit.org/show_bug.cgi?id=291288</a>
<a href="https://rdar.apple.com/148843842">rdar://148843842</a>

Reviewed by Wenson Hsieh and Ada Chan.

This PR adds an extra IPC call between the web process and the model process when we drag a model
element, to animate the model during the drag. This is meant to position and transform the model
such that the snapshot we retrieve is not in a bad state.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::animateModelToFitPortal):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::animateModelToFitPortal):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::tryToBeginDragAtPoint):
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::animateModelToFitPortal):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didAnimateModelForDrag):
(WebKit::ModelProcessModelPlayer::animateModelToFitPortal):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didAnimateModelForDrag):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/294549@main">https://commits.webkit.org/294549@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78bb61c9c264ff856b82a801f1ec21ad6cd3677a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102268 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21936 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12251 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107428 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/52904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104307 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30443 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77815 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/52904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105274 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17215 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92318 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58153 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17049 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10346 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52262 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86875 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10416 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109803 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29400 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21672 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86797 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29764 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88521 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86385 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21970 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31195 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8913 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23642 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29328 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34623 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29139 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32462 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30698 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -535,6 +535,12 @@ void HTMLModelElement::renderingAbruptlyStopped()
+void HTMLModelElement::animateModelToFitPortal()
+{
+    if (hasPortal() && m_modelPlayer)
+        m_modelPlayer->animateModelToFitPortal();
```
I saw that the implementation of `ModelPlayer::animateModelToFitPortal()` is guarded with `ENABLE(DRAG_SUPPORT)`, but the caller and the declaration of that method are not. Let's keep those consistent.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
```diff
@@ -42,6 +42,7 @@ messages -> ModelProcessModelPlayerProxy {
     BeginStageModeTransform(WebCore::TransformationMatrix transform)
     UpdateStageModeTransform(WebCore::TransformationMatrix transform)
     EndStageModeInteraction()
+    AnimateModelToFitPortal()
```
Ditto about the `ENABLE(DRAG_SUPPORT)` guard.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -756,6 +756,11 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+void ModelProcessModelPlayerProxy::animateModelToFitPortal()
+{
+    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=291289
+}
```
Ditto about the `ENABLE(DRAG_SUPPORT)` guard.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -762,6 +762,11 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
 
+void ModelProcessModelPlayerProxy::animateModelToFitPortal()
+{
+    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=291289
```
Until we address the FIXME, maybe this should send the `ModelProcessModelPlayer.DidAnimateModelForDrag` message back to the web content process for the time being.

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5590,6 +5590,12 @@ void WebPage::stageModeSessionDidEnd(std::optional<ElementIdentifier> elementID)
+void WebPage::didAnimateModelForDrag()
+{
+    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=291284
+    send(Messages::WebPageProxy::DidHandleDragStartRequest(WebCore::DragStartRequestResult::Started));
```
```suggestion
    send(Messages::WebPageProxy::DidHandleDragStartRequest(DragStartRequestResult::Started));
```

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -528,6 +528,12 @@ void HTMLModelElement::renderingAbruptlyStopped()
+void HTMLModelElement::animateModelToFitPortal()
+{
+    if (hasPortal() && m_modelPlayer)
+        m_modelPlayer->animateModelToFitPortal();
```
Just realized we can have a case where `HTMLModelElement::animateModelToFitPortal()` doesn't do anything: e.g. if there's no portal, or there's no model player, or the model player does not handle animating model to fit portal. In that case, we'd still end up returning a delayed result and the drag request will be stuck.

Maybe this method could be `tryAnimateModelToFitPortal()` that returns a boolean instead. It'll return true only if the animation will happen. Then in `EventHandler::tryToBeginDragAtPoint()`, we'll only return the delayed result if `tryAnimateModelToFitPortal()` returns true.

achan00 commented on Source/WebCore/Modules/model-element/ModelPlayer.h:
```diff
@@ -102,6 +102,7 @@ class WEBCORE_EXPORT ModelPlayer : public RefCounted<ModelPlayer> {
     virtual void updateStageModeTransform(const TransformationMatrix&);
     virtual void endStageModeInteraction();
     virtual void renderingAbruptlyStopped();
+    virtual void animateModelToFitPortal();
```
This should probably return a `bool` too, and the default implementation would return `false`.

Phipson committed:
Add call to animate model to center of portal on drag
https://bugs.webkit.org/show_bug.cgi?id=291288
rdar://148843842

Reviewed by Wenson Hsieh and Ada Chan.

This PR adds an extra IPC call between the web process and the model process when we drag a model
element, to animate the model during the drag. This is meant to position and transform the model
such that the snapshot we retrieve is not in a bad state.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::animateModelToFitPortal):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::animateModelToFitPortal):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::tryToBeginDragAtPoint):
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::animateModelToFitPortal):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didAnimateModelForDrag):
(WebKit::ModelProcessModelPlayer::animateModelToFitPortal):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didAnimateModelForDrag):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/294549@main
